api.go

```diff
@@ -16,43 +16,112 @@
+// WithAdditionalMiddleware adds the specified middleware *after* the default middleware.
+// Can be called multiple times.
+func WithAdditionalMiddleware(m middlewareFunc) Option {
+	return func(c *config) {
+		c.additionalMiddleware = append(c.additionalMiddleware, m)
+	}
 }
 
-func AttachRoutes(router chi.Router, serviceBroker ServiceBroker, logger *slog.Logger) {
-	attachRoutes(router, serviceBroker, logger)
+func WithOptions(opts ...Option) Option {
+	return func(c *config) {
+		for _, o := range opts {
+			o(c)
+		}
+	}
 }
 
-func attachRoutes(router chi.Router, serviceBroker ServiceBroker, logger *slog.Logger) {
+func router(serviceBroker ServiceBroker, logger *slog.Logger) http.Handler {
 	apiHandler := handlers.NewApiHandler(serviceBroker, logger)
-	router.Get("/v2/catalog", apiHandler.Catalog)
+	r := http.NewServeMux()
+	r.HandleFunc("GET /v2/catalog", apiHandler.Catalog)
+
+	r.HandleFunc("PUT /v2/service_instances/{instance_id}", apiHandler.Provision)
+	r.HandleFunc("GET /v2/service_instances/{instance_id}", apiHandler.GetInstance)
+	r.HandleFunc("PATCH /v2/service_instances/{instance_id}", apiHandler.Update)
+	r.HandleFunc("DELETE /v2/service_instances/{instance_id}", apiHandler.Deprovision)
+
+	r.HandleFunc("GET /v2/service_instances/{instance_id}/last_operation", apiHandler.LastOperation)
 
-	router.Get("/v2/service_instances/{instance_id}", apiHandler.GetInstance)
-	router.Put("/v2/service_instances/{instance_id}", apiHandler.Provision)
-	router.Delete("/v2/service_instances/{instance_id}", apiHandler.Deprovision)
-	router.Get("/v2/service_instances/{instance_id}/last_operation", apiHandler.LastOperation)
-	router.Patch("/v2/service_instances/{instance_id}", apiHandler.Update)
+	r.HandleFunc("PUT /v2/service_instances/{instance_id}/service_bindings/{binding_id}", apiHandler.Bind)
+	r.HandleFunc("GET /v2/service_instances/{instance_id}/service_bindings/{binding_id}", apiHandler.GetBinding)
+	r.HandleFunc("DELETE /v2/service_instances/{instance_id}/service_bindings/{binding_id}", apiHandler.Unbind)
+
+	r.HandleFunc("GET /v2/service_instances/{instance_id}/service_bindings/{binding_id}/last_operation", apiHandler.LastBindingOperation)
+
+	return r
+}
+
+func defaultMiddleware(logger *slog.Logger) []middlewareFunc {
+	return []middlewareFunc{
+		middlewares.APIVersionMiddleware{Logger: logger}.ValidateAPIVersionHdr,
+		middlewares.AddCorrelationIDToContext,
+		middlewares.AddOriginatingIdentityToContext,
+		middlewares.AddInfoLocationToContext,
+		middlewares.AddRequestIdentityToContext,
+	}
+}
 
-	router.Get("/v2/service_instances/{instance_id}/service_bindings/{binding_id}", apiHandler.GetBinding)
-	router.Put("/v2/service_instances/{instance_id}/service_bindings/{binding_id}", apiHandler.Bind)
-	router.Delete("/v2/service_instances/{instance_id}/service_bindings/{binding_id}", apiHandler.Unbind)
+func combineMiddlewares(middlewares ...middlewareFunc) middlewareFunc {
+	reversed := middlewares[:]
+	slices.Reverse(reversed)
```

Comment:
question:

line 45, the order of middleware being passed:
1. `cfg.authMiddleware` (first)
2. `defaultMiddleware(logger)` (middle)
3. `cfg.additionalMiddleware` (last)

Looking at the intended order of execution from the code comments:
* `WithCustomAuth`: "adds the specified middleware before any other middleware"
* `WithAdditionalMiddleware`: "adds the specified middleware after the default middleware"

The current order in the slice already matches our desired execution order:
1. Auth middleware should run first
2. Default middleware second
3. Additional middleware last

Am I wrong?

Reply:
You are correct about the order, and in order to achieve this we need to reverse the slice. I've moved this to another package and added tests and a comment to make this clear (including to myself!)